src/git.ts

```diff
@@ -51,56 +37,72 @@ async function gitRemoteURL(repoDirectory: string, remoteName: string): Promise<
+ *
+ * @param repoDirectory the repository root directory
  */
-async function gitDefaultRemoteURL(repoDirectory: string): Promise<string> {
-    const [remote] = await gitRemoteBranch(repoDirectory)
-    if (remote !== '') {
-        return gitRemoteURL(repoDirectory, remote)
+async function gitRemoteNameAndBranch(repoDirectory: string): Promise<RemoteName & Branch> {
+    let remoteName = ''
+    let branch = 'HEAD'
+
+    const { stdout } = await execa('git', ['rev-parse', '--abbrev-ref', 'HEAD@{upstream}'], { cwd: repoDirectory })
+    const remoteAndBranch = stdout.split('/')
+
+    if (remoteAndBranch.length === 1) {
+        // The upstream branch points to a local branch.
+        ;[remoteName] = remoteAndBranch
+    }
+    if (remoteAndBranch.length === 2) {
+        ;[remoteName, branch] = remoteAndBranch
     }
 
     // If we cannot find the remote name deterministically, we use the first
     // Git remote found.
-    const remotes = await gitRemotes(repoDirectory)
-    if (remotes.length === 0) {
-        throw new Error('no configured git remotes')
-    }
-    if (remotes.length > 1) {
-        log.appendLine(`using first git remote: ${remotes[0]}`)
+    if (!remoteName) {
+        const remotes = await gitRemotes(repoDirectory)
+        if (remotes.length === 0) {
+            throw new Error('no configured git remotes')
+        }
+        if (remotes.length > 1) {
+            log.appendLine(`using first git remote: ${remotes[0]}`)
+            remoteName = remotes[0]
+        }
     }
-    return gitRemoteURL(repoDirectory, remotes[0])
-}
 
-/**
- * Returns the repository root directory for any directory within the
- * repository.
- */
-async function gitRootDirectory(repoDirectory: string): Promise<string> {
-    const { stdout } = await execa('git', ['rev-parse', '--show-toplevel'], { cwd: repoDirectory })
-    return stdout
+    return { remoteName, branch }
 }
 
-/**
- * Returns either the current remote branch name of the repository OR in all
- * other cases (e.g. detached HEAD state, upstream branch points to a local
- * branch), it returns "HEAD".
- */
-async function gitBranch(repoDirectory: string): Promise<string> {
-    const [origin, branch] = await gitRemoteBranch(repoDirectory)
-    if (origin !== '') {
-        // The remote branch exists.
-        return branch
-    }
-    return 'HEAD'
+interface RepositoryInfo extends Branch {
+    /** Git repository remote URL */
+    remoteURL: string
+
+    /** File path relative to the repository root */
+    fileRelative: string
 }
 
 /**
  * Returns the Git repository remote URL, the current branch, and the file path
  * relative to the repository root. Empty strings are returned if this cannot be
  * determined.
  */
-export async function repoInfo(filePath: string): Promise<[string, string, string]> {
+export async function repoInfo(filePath: string): Promise<RepositoryInfo> {
     let remoteURL = ''
     let branch = ''
     let fileRelative = ''
```

Comment:
Same comment here

Reply:
It looks like we do this to "salvage" variables already re-assigned by the time an error is thrown. The issue is that whether we use the other values is dependent on whether `remoteURL` is found. Is there ever a case where we'd catch an error but still end up with `remoteURL`? I can't think of any since it's the last value we look for.